// graphiql/dist/components/ExecuteButton.js
/**
 *  Copyright (c) 2015, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the license found in the
 *  LICENSE-examples file in the root directory of this source tree.
 */

"use strict";

exports.__esModule = true;

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

/**
 * ExecuteButton
 *
 * What a nice round shiny button. Cmd/Ctrl-Enter is the shortcut.
 */

var ExecuteButton = (function (_React$Component) {
  _inherits(ExecuteButton, _React$Component);

  function ExecuteButton() {
    _classCallCheck(this, ExecuteButton);

    _React$Component.apply(this, arguments);
  }

  ExecuteButton.prototype.render = function render() {
    return _react2["default"].createElement(
      "button",
      {
        className: "execute-button",
        onClick: this.props.onClick,
        title: "Execute Query (Ctrl-Enter)" },
      _react2["default"].createElement(
        "svg",
        { width: "34", height: "34" },
        _react2["default"].createElement("path", { d: "M 11 9 L 24 16 L 11 23 z" })
      )
    );
  };

  ExecuteButton.prototype.componentDidMount = function componentDidMount() {
    var _this = this;

    this.keyHandler = function (event) {
      if ((event.metaKey || event.ctrlKey) && event.keyCode === 13) {
        event.preventDefault();
        if (_this.props.onClick) {
          _this.props.onClick();
        }
      }
    };
    document.addEventListener('keydown', this.keyHandler, true);
  };

  ExecuteButton.prototype.componentWillUnmount = function componentWillUnmount() {
    document.removeEventListener('keydown', this.keyHandler, true);
  };

  _createClass(ExecuteButton, null, [{
    key: "propTypes",
    value: {
      onClick: _react.PropTypes.func
    },
    enumerable: true
  }]);

  return ExecuteButton;
})(_react2["default"].Component);

exports.ExecuteButton = ExecuteButton;
