// graphiql/dist/components/VariableEditor.js
/**
 *  Copyright (c) 2015, Facebook, Inc.
 *  All rights reserved.
 *
 *  This source code is licensed under the license found in the
 *  LICENSE-examples file in the root directory of this source tree.
 */

'use strict';

exports.__esModule = true;

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _reactDom2 = _interopRequireDefault(_reactDom);

var _codemirror = require('codemirror');

var _codemirror2 = _interopRequireDefault(_codemirror);

require('codemirror/addon/fold/brace-fold');

require('codemirror/addon/fold/foldgutter');

require('codemirror/addon/lint/lint');

require('codemirror/keymap/sublime');

require('codemirror/mode/javascript/javascript');

require('../codemirror/lint/json-lint');

/**
 * VariableEditor
 *
 * An instance of CodeMirror for editing variables defined in QueryEditor.
 *
 * Props:
 *
 *   - value: The text of the editor.
 *   - onEdit: A function called when the editor changes, given the edited text.
 *
 */

var VariableEditor = (function (_React$Component) {
  _inherits(VariableEditor, _React$Component);

  _createClass(VariableEditor, null, [{
    key: 'propTypes',
    value: {
      value: _react.PropTypes.string,
      onEdit: _react.PropTypes.func
    },
    enumerable: true
  }]);

  function VariableEditor(props) {
    var _this = this;

    _classCallCheck(this, VariableEditor);

    _React$Component.call(this);

    // Keep a cached version of the value, this cache will be updated when the
    // editor is updated, which can later be used to protect the editor from
    // unnecessary updates during the update lifecycle.

    this._onEdit = function () {
      if (!_this.ignoreChangeEvent) {
        _this.cachedValue = _this.editor.getValue();
        if (_this.props.onEdit) {
          _this.props.onEdit(_this.cachedValue);
        }
      }
    };

    this.cachedValue = props.value || '';
  }

  VariableEditor.prototype.componentDidMount = function componentDidMount() {
    this.editor = _codemirror2['default'](_reactDom2['default'].findDOMNode(this), {
      value: this.props.value || '',
      lineNumbers: true,
      theme: 'graphiql',
      mode: {
        name: 'javascript',
        json: true
      },
      lint: true,
      autoCloseBrackets: true,
      matchBrackets: true,
      showCursorWhenSelecting: true,
      keyMap: 'sublime',
      foldGutter: {
        minFoldSize: 4
      },
      gutters: ['CodeMirror-linenumbers', 'CodeMirror-foldgutter'],
      extraKeys: {
        // Editor improvements
        'Ctrl-Left': 'goSubwordLeft',
        'Ctrl-Right': 'goSubwordRight',
        'Alt-Left': 'goGroupLeft',
        'Alt-Right': 'goGroupRight'
      }
    });

    this.editor.on('change', this._onEdit);
  };

  VariableEditor.prototype.componentWillUnmount = function componentWillUnmount() {
    this.editor.off('change', this._onEdit);
    this.editor = null;
  };

  VariableEditor.prototype.componentDidUpdate = function componentDidUpdate(prevProps) {
    // Ensure the changes caused by this update are not interpretted as
    // user-input changes which could otherwise result in an infinite
    // event loop.
    this.ignoreChangeEvent = true;
    if (this.props.value !== prevProps.value && this.props.value !== this.cachedValue) {
      this.cachedValue = this.props.value;
      this.editor.setValue(this.props.value);
    }
    this.ignoreChangeEvent = false;
  };

  VariableEditor.prototype.render = function render() {
    return _react2['default'].createElement('div', { className: 'codemirrorWrap', ref: 'codemirror' });
  };

  return VariableEditor;
})(_react2['default'].Component);

exports.VariableEditor = VariableEditor;
